File: porsh/src/test/java/eu/darken/porter/porsh/FakeHost.java
package eu.darken.porter.porsh;

import java.util.concurrent.CountDownLatch;

/**
 * A {@link PorshHost} that never forks. {@code start()} is the only member that reaches
 * JNI, so overriding it is enough to exercise the registry off-device.
 */
class FakeHost extends PorshHost {

    private final int pid;

    /** Opens once a thread has entered awaitExitCode on this host. */
    final CountDownLatch awaitEntered = new CountDownLatch(1);

    volatile long windowSize = -1;
    volatile long exitedAt;

    FakeHost(int pid) {
        super(new String[0], null, null, (byte) 0, null, null, null);
        this.pid = pid;
    }

    @Override
    public void start() {
    }

    @Override
    public int getPid() {
        return pid;
    }

    @Override
    public void setWindowSize(long size) {
        windowSize = size;
    }

    @Override
    long getExitedAtMillis() {
        return exitedAt;
    }

    @Override
    int awaitExitCode(long timeoutMillis) {
        awaitEntered.countDown();
        return super.awaitExitCode(timeoutMillis);
    }
}

File: porsh/src/test/java/eu/darken/porter/porsh/FakeHost.kt
package eu.darken.porter.porsh

import java.util.concurrent.CountDownLatch

/**
 * A [PorshHost] that never forks. `start()` is the only member that reaches JNI, so
 * overriding it is enough to exercise the registry off-device.
 */
internal open class FakeHost(override val pid: Int) : PorshHost(emptyArray(), null, null, 0, null, null, null) {

    /** Opens once a thread has entered awaitExitCode on this host. */
    val awaitEntered = CountDownLatch(1)

    @Volatile
    var lastWindowSize = -1L

    @Volatile
    var exitedAt = 0L

    override fun start() {
    }

    override fun setWindowSize(size: Long) {
        lastWindowSize = size
    }

    override val exitedAtMillis: Long
        get() = exitedAt

    override fun awaitExitCode(timeoutMillis: Long): Int {
        awaitEntered.countDown()
        return super.awaitExitCode(timeoutMillis)
    }
}
